csharp/src/Apache.Arrow/DecimalUtility.cs

```diff
@@ -0,0 +1,188 @@
+            } else
+            {
+                Buffer.BlockCopy(integerBytes, 0, result, 0, integerBytes.Length);
+            }
+
+            return result;
+        }
+
+        public static bool CheckPrecisionAndScale(decimal value, int precision, int scale, out BigInteger integerValue)
+        {
+            DecimalLayout layout = new DecimalLayout(value); // use in place of decimal.GetBits(value) to avoid an allocation
+            integerValue = new BigInteger(BitConverter.GetBytes(layout.Lo).Concat(BitConverter.GetBytes(layout.Mid)).Concat(BitConverter.GetBytes(layout.Hi)).ToArray());
+
+            if (layout.Scale > scale)
+                throw new OverflowException("Decimal scale can not be greater than that in the Arrow vector: " + layout.Scale + " != " + scale);
+
+            if(integerValue >= BigInteger.Pow(10, precision))
+                throw new OverflowException("Decimal precision can not be greater than that in the Arrow vector: " + value + " has precision > " + precision);
+
+            if (layout.Scale < scale) // pad with trailing zeros
+            {
+                integerValue *= BigInteger.Pow(10, scale - layout.Scale);
+            }
+
+            if (value < 0) // sign the big int
+                integerValue = -integerValue;
+
+            return true;
+        }
+
+        private static decimal GetDecimalFromBigInteger(BigInteger value, int scale)
+        {
+            var b = new BigIntegerLayout(value);
+            if (b.Bits == null)
+                return b.Sign;
+
+            int length = b.Bits.Length;
+            if (length > 3) throw new OverflowException("Decimal overflow");
+
+            int lo = 0, mi = 0, hi = 0;
+
+            unchecked
+            {
+                if (length > 2) hi = b.Bits[2];
+                if (length > 1) mi = b.Bits[1];
+                if (length > 0) lo = b.Bits[0];
+            }
+
+            return new decimal(lo, mi, hi, b.Sign < 0, (byte)scale);
+        }
+
+        [StructLayout(LayoutKind.Explicit)]
+        readonly struct BigIntegerLayout
+        {
+            public BigIntegerLayout(BigInteger value)
+            {
+                this = default;
+                bi = value;
+            }
+
+            [FieldOffset(0)] readonly BigInteger bi;
+            [FieldOffset(0)] readonly int[] bits;
+
+            public int Sign => bi.Sign;
+            public int[] Bits =>  bits;
+        }
+
+
+        [StructLayout(LayoutKind.Explicit)]
+        readonly struct DecimalLayout
+        {
+            public DecimalLayout(decimal value)
+            {
+                this = default;
+                d = value;
+            }
+
+            [FieldOffset(0)] readonly decimal d;
+            [FieldOffset(0)] readonly int flags;
+            [FieldOffset(4)] readonly int hi;
+            [FieldOffset(8)] readonly int lo;
+            [FieldOffset(12)] readonly int mid;
```

Comment:
This isn't exactly how System.Decimal is defined. Would this work on big endian systems?

https://github.com/dotnet/runtime/blob/8a52f1e948b6f22f418817ec1068f07b8dae2aa5/src/libraries/System.Private.CoreLib/src/System/Decimal.cs#L106-L108

```
        private readonly int _flags;
        private readonly uint _hi32;
        private readonly ulong _lo64;
```

Reply:
Updated to support both big and little endian systems, accessing lo / mid.
Agreed it doesn't match exactly how Decimal is defined but I have since found that this is how DecCalc accesses: https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Decimal.DecCalc.cs#L40